soccer/planning/planner/CollectPlanner.cpp

```diff
@@ -0,0 +1,139 @@
+    RobotInstant startInstant = request.start;
+    RotationConstraints rotationConstraints = request.constraints.rot;
+    Point targetPoint =
+        ball.pos - approachDir * (Ball_Radius + Robot_MouthRadius);
+    request.constraints.mot.maxSpeed =
+        ball.vel.dot(approachDir) + *_touchDeltaSpeed;
+    Point targetVel = request.constraints.mot.maxSpeed * approachDir;
+    RobotInstant targetInstant{Pose{targetPoint, 0}, Twist{targetVel, 0},
+                               RJ::Time{0s}};
+    request.motionCommand = PathTargetCommand{targetInstant};
+    Trajectory path = _pathTargetPlanner.plan(std::move(request));
+    PlanAngles(path, startInstant, AngleFns::faceAngle(approachDir.angle()),
+               rotationConstraints);
+    return std::move(path);
+}
+Trajectory CollectPlanner::control(PlanRequest&& request) {
+    Point approachDir = findApproachDirection(request);
+    RotationConstraints rotationConstraints = request.constraints.rot;
+    RobotInstant startInstant = request.start;
+    // move with constant negative acceleration until stop
+    double accel =
+        request.constraints.mot.maxAcceleration * *_controlAccelScalePercent;
+    double stoppingDist =
+        std::pow(startInstant.velocity.linear().mag(), 2) / (2 * accel);
+    Point targetPoint =
+        startInstant.pose.position() + stoppingDist * approachDir;
+    RobotInstant targetInstant{Pose{targetPoint, 0}, {}, RJ::Time{0s}};
+    request.motionCommand = PathTargetCommand{targetInstant};
+    request.constraints.mot.maxSpeed = std::min(
+        startInstant.velocity.linear().mag(), request.constraints.mot.maxSpeed);
+    request.constraints.mot.maxAcceleration = accel;
+    Trajectory path = _pathTargetPlanner.plan(std::move(request));
+    PlanAngles(path, startInstant, AngleFns::faceAngle(approachDir.angle()),
+               rotationConstraints);
+    return std::move(path);
+}
+void CollectPlanner::processStateTransitions(const PlanRequest& request) {
+    const Ball& ball = request.context->state.ball;
+    RobotInstant startInstant = request.start;
+    double distToBall = (startInstant.pose.position() - ball.pos).mag();
+    double distToBallAtContact = Ball_Radius + Robot_MouthRadius;
+    CollectState& collectState = _collectStates[request.shellID];
+    const double distCutoffToFine = *_bufferDistBeforeContact + 0.0765;
+    // course -> fine
+    if (collectState == CollectState::Course) {
+        if (distToBall < distCutoffToFine + distToBallAtContact) {
+            collectState = CollectState::Fine;
+        }
+    }
+    // fine -> control
+    if (collectState == CollectState::Fine) {
+        if (distToBall < *_distCutoffToControl + distToBallAtContact) {
+            collectState = CollectState::Control;
+        }
+    }
+    // backward transitions
+    bool fineToCourse =
+        collectState == CollectState::Fine &&
+        distToBall > distCutoffToFine + distToBallAtContact + 0.2;
+    bool controlToCourse =
```

Comment:
If you have access to robot heading, a good check would be to see if the ball is not in the mouth or inside the robot. If the robot misses on the final approach, it's possible that the distance trigger will not happen. (Make sure to account for our robot blocking vision updates as well since that's a significant cause of problems close up